File: ts/src/model-catalog.ts
import { ChatGPTOAuthError } from "./auth.js";

export interface ReasoningLevel {
  effort: string;
  description?: string;
}

export interface ModelCatalogEntry {
  slug: string;
  displayName: string;
  description?: string;
  defaultReasoningEffort?: string;
  supportedReasoningLevels: ReasoningLevel[];
  contextWindow?: number;
  maxContextWindow?: number;
  supportedInApi?: boolean;
  capabilities: Record<string, unknown>;
}

export interface ResolvedModel {
  requestedModel: string;
  upstreamModel: string;
  reasoningEffort?: string;
  catalogEntry?: ModelCatalogEntry;
  alias: boolean;
}

const MODEL_LIST_KEYS = ["models", "data"] as const;
const LUNA_MODEL_SLUG = "gpt-5.6-luna";
const LUNA_ALIAS_PREFIX = "luna-";

export function isLunaModelId(model: string): boolean {
  return model === LUNA_MODEL_SLUG
    || model.startsWith(`${LUNA_MODEL_SLUG}-`)
    || model.startsWith(LUNA_ALIAS_PREFIX);
}

export function normalizeModelCatalog(value: unknown): ModelCatalogEntry[] {
  const rawModels = Array.isArray(value)
    ? value
    : isRecord(value)
      ? MODEL_LIST_KEYS
        .map((key) => value[key])
        .find((candidate): candidate is unknown[] => Array.isArray(candidate))
      : undefined;
  if (!Array.isArray(rawModels)) {
    throw new ChatGPTOAuthError("Codex model catalog response is missing a models array");
  }

  const entries: ModelCatalogEntry[] = [];
  for (const [index, raw] of rawModels.entries()) {
    if (!isRecord(raw)) continue;
    const slug = stringValue(raw.slug ?? raw.model ?? raw.id);
    if (!slug) {
      throw new ChatGPTOAuthError(`Codex model catalog entry ${index} is missing a slug`);
    }
    const supported = normalizeReasoningLevels(
      raw.supported_reasoning_levels
        ?? raw.supported_reasoning_efforts
        ?? raw.reasoning_levels,
    );
    const defaultReasoningEffort = stringValue(
      raw.default_reasoning_level
        ?? raw.default_reasoning_effort
        ?? raw.default_reasoning,
    );
    if (
      defaultReasoningEffort != null
      && supported.length > 0
      && !supported.some((level) => level.effort === defaultReasoningEffort)
    ) {
      throw new ChatGPTOAuthError(
        `Codex model catalog entry ${JSON.stringify(slug)} has an unsupported default reasoning level`,
      );
    }
    entries.push({
      slug,
      displayName: stringValue(raw.display_name ?? raw.displayName ?? raw.name) ?? slug,
      description: stringValue(raw.description),
      defaultReasoningEffort,
      supportedReasoningLevels: supported,
      contextWindow: positiveInteger(raw.context_window),
      maxContextWindow: positiveInteger(raw.max_context_window),
      supportedInApi: typeof raw.supported_in_api === "boolean"
        ? raw.supported_in_api
        : typeof raw.supportedInApi === "boolean"
          ? raw.supportedInApi
          : undefined,
      capabilities: { ...raw },
    });
  }
  return entries;
}

export function resolveModelAlias(
  requestedModel: string,
  catalog: readonly ModelCatalogEntry[],
): ResolvedModel {
  const exact = catalog.find((entry) => entry.slug === requestedModel);
  if (exact) {
    return {
      requestedModel,
      upstreamModel: exact.slug,
      catalogEntry: exact,
      alias: false,
    };
  }

  // Only strip a suffix when the complete base model is present and the suffix
  // is advertised by that model. Unknown IDs therefore remain untouched.
  for (const entry of catalog) {
    const prefix = `${entry.slug}-`;
    if (!requestedModel.startsWith(prefix)) continue;
    const effort = requestedModel.slice(prefix.length);
    if (!effort || !entry.supportedReasoningLevels.some((level) => level.effort === effort)) {
      continue;
    }
    return {
      requestedModel,
      upstreamModel: entry.slug,
      reasoningEffort: effort,
      catalogEntry: entry,
      alias: true,
    };
  }

  if (requestedModel.startsWith(LUNA_ALIAS_PREFIX)) {
    const luna = catalog.find((entry) => entry.slug === LUNA_MODEL_SLUG);
    const effort = requestedModel.slice(LUNA_ALIAS_PREFIX.length);
    if (
      luna != null
      && effort
      && luna.supportedReasoningLevels.some((level) => level.effort === effort)
    ) {
      return {
        requestedModel,
        upstreamModel: luna.slug,
        reasoningEffort: effort,
        catalogEntry: luna,
        alias: true,
      };
    }
  }

  return {
    requestedModel,
    upstreamModel: requestedModel,
    alias: false,
  };
}

export function publicModelsFromCatalog(
  catalog: readonly ModelCatalogEntry[],
  now = Math.floor(Date.now() / 1000),
): Record<string, unknown>[] {
  const models: Record<string, unknown>[] = [];
  for (const entry of catalog) {
    if (entry.supportedInApi === false) continue;
    const virtualAliases: string[] = [];
    const add = (id: string, suffix?: string): void => {
      models.push({
        id,
        object: "model",
        created: now,
        owned_by: "openai",
        display_name: suffix == null
          ? entry.displayName
          : `${entry.displayName} (${suffix})`,
        description: entry.description,
        context_window: entry.contextWindow ?? entry.maxContextWindow,
        default_reasoning_effort: suffix ?? entry.defaultReasoningEffort,
      });
    };
    add(entry.slug);
    for (const level of entry.supportedReasoningLevels) {
      add(`${entry.slug}-${level.effort}`, level.effort);
      if (entry.slug === LUNA_MODEL_SLUG) {
        virtualAliases.push(level.effort);
      }
    }
    for (const effort of virtualAliases) {
      add(`${LUNA_ALIAS_PREFIX}${effort}`, effort);
    }
  }
  return models;
}

function normalizeReasoningLevels(value: unknown): ReasoningLevel[] {
  if (!Array.isArray(value)) return [];
  const levels: ReasoningLevel[] = [];
  const seen = new Set<string>();
  for (const raw of value) {
    const effort = typeof raw === "string"
      ? raw
      : isRecord(raw)
        ? stringValue(raw.effort ?? raw.level ?? raw.name)
        : undefined;
    if (!effort || seen.has(effort)) continue;
    seen.add(effort);
    levels.push({
      effort,
      description: isRecord(raw) ? stringValue(raw.description) : undefined,
    });
  }
  return levels;
}

function stringValue(value: unknown): string | undefined {
  return typeof value === "string" && value.trim().length > 0 ? value : undefined;
}

function positiveInteger(value: unknown): number | undefined {
  return typeof value === "number" && Number.isSafeInteger(value) && value > 0
    ? value
    : undefined;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: src/codex_as_api/model_catalog.py
from __future__ import annotations

import dataclasses
from typing import Any


@dataclasses.dataclass(frozen=True, slots=True)
class ReasoningLevel:
    effort: str
    description: str | None = None


@dataclasses.dataclass(frozen=True, slots=True)
class ModelCatalogEntry:
    slug: str
    display_name: str
    description: str | None
    default_reasoning_effort: str | None
    supported_reasoning_levels: tuple[ReasoningLevel, ...]
    context_window: int | None
    max_context_window: int | None
    supported_in_api: bool | None
    capabilities: dict[str, Any]


@dataclasses.dataclass(frozen=True, slots=True)
class ResolvedModel:
    requested_model: str
    upstream_model: str
    reasoning_effort: str | None
    catalog_entry: ModelCatalogEntry | None
    alias: bool


def normalize_model_catalog(value: object) -> list[ModelCatalogEntry]:
    if isinstance(value, list):
        raw_models = value
    elif isinstance(value, dict):
        candidates = [value.get(key) for key in ("models", "data")]
        raw_models = next((candidate for candidate in candidates if isinstance(candidate, list)), None)
    else:
        raw_models = None
    if not isinstance(raw_models, list):
        raise ValueError("Codex model catalog response is missing a models array")

    entries: list[ModelCatalogEntry] = []
    for index, raw in enumerate(raw_models):
        if not isinstance(raw, dict):
            continue
        slug = _string(raw.get("slug") or raw.get("model") or raw.get("id"))
        if slug is None:
            raise ValueError(f"Codex model catalog entry {index} is missing a slug")
        levels = _reasoning_levels(
            raw.get("supported_reasoning_levels")
            or raw.get("supported_reasoning_efforts")
            or raw.get("reasoning_levels")
        )
        default_effort = _string(
            raw.get("default_reasoning_level")
            or raw.get("default_reasoning_effort")
            or raw.get("default_reasoning")
        )
        if default_effort is not None and levels and default_effort not in {level.effort for level in levels}:
            raise ValueError(f"Codex model catalog entry {slug!r} has an unsupported default reasoning level")
        entries.append(
            ModelCatalogEntry(
                slug=slug,
                display_name=_string(raw.get("display_name") or raw.get("displayName") or raw.get("name")) or slug,
                description=_string(raw.get("description")),
                default_reasoning_effort=default_effort,
                supported_reasoning_levels=tuple(levels),
                context_window=_positive_int(raw.get("context_window")),
                max_context_window=_positive_int(raw.get("max_context_window")),
                supported_in_api=(
                    raw.get("supported_in_api")
                    if isinstance(raw.get("supported_in_api"), bool)
                    else raw.get("supportedInApi")
                    if isinstance(raw.get("supportedInApi"), bool)
                    else None
                ),
                capabilities=dict(raw),
            )
        )
    return entries


def resolve_model_alias(requested_model: str, catalog: list[ModelCatalogEntry]) -> ResolvedModel:
    exact = next((entry for entry in catalog if entry.slug == requested_model), None)
    if exact is not None:
        return ResolvedModel(requested_model, exact.slug, None, exact, False)
    for entry in catalog:
        prefix = f"{entry.slug}-"
        if not requested_model.startswith(prefix):
            continue
        effort = requested_model[len(prefix) :]
        if effort and any(level.effort == effort for level in entry.supported_reasoning_levels):
            return ResolvedModel(requested_model, entry.slug, effort, entry, True)
    return ResolvedModel(requested_model, requested_model, None, None, False)


def public_models_from_catalog(catalog: list[ModelCatalogEntry], created: int) -> list[dict[str, Any]]:
    models: list[dict[str, Any]] = []
    for entry in catalog:
        if entry.supported_in_api is False:
            continue
        models.append(_public_model(entry, entry.slug, created, None))
        for level in entry.supported_reasoning_levels:
            models.append(_public_model(entry, f"{entry.slug}-{level.effort}", created, level.effort))
    return models


def _public_model(entry: ModelCatalogEntry, model_id: str, created: int, effort: str | None) -> dict[str, Any]:
    return {
        "id": model_id,
        "object": "model",
        "created": created,
        "owned_by": "openai",
        "display_name": entry.display_name if effort is None else f"{entry.display_name} ({effort})",
        "description": entry.description,
        "context_window": entry.context_window or entry.max_context_window,
        "default_reasoning_effort": effort or entry.default_reasoning_effort,
    }


def _reasoning_levels(value: object) -> list[ReasoningLevel]:
    if not isinstance(value, list):
        return []
    levels: list[ReasoningLevel] = []
    seen: set[str] = set()
    for raw in value:
        if isinstance(raw, str):
            effort, description = raw, None
        elif isinstance(raw, dict):
            effort = _string(raw.get("effort") or raw.get("level") or raw.get("name"))
            description = _string(raw.get("description"))
        else:
            continue
        if effort is None or effort in seen:
            continue
        seen.add(effort)
        levels.append(ReasoningLevel(effort, description))
    return levels


def _string(value: object) -> str | None:
    return value if isinstance(value, str) and value.strip() else None


def _positive_int(value: object) -> int | None:
    return value if isinstance(value, int) and not isinstance(value, bool) and value > 0 else None
